File: Preprocessing.Cs/Preprocessing/GraphGenerator.cs
using System.Text.RegularExpressions;
using Microsoft.CodeAnalysis.CSharp;

namespace AstGenerator;

public class GraphGenerator
{
    public static IEnumerable<Graph> GetGraphs(string fileName, string code)
    {
        var tree = CSharpSyntaxTree.ParseText(code);
        var root = tree.GetRoot();

        var methodVisitor = new MethodDeclarationVisitor();
        methodVisitor.Visit(root);

        var methodDeclarations = methodVisitor.MethodDeclarations;

        foreach (var method in methodDeclarations)
        {
            var methodDeclaration = method.Declaration;

            var className = method.ClassDeclaration?.Identifier.ToString();
            var methodName = methodDeclaration.Identifier.ToString();
            var fullName = className == null ? methodName : $"{className}->{methodName}";

            var lineNumber = methodDeclaration.GetLocation().GetLineSpan().StartLinePosition.Line;

            var lines = ConvertToLines(code);

            var graphVisitor = new AstVisitor();
            graphVisitor.Process(methodDeclaration);

            yield return new Graph(fileName, "unknown", fullName, lineNumber, 0,
                lines, graphVisitor.GetNodes(),
                graphVisitor.GetAstEdges(), graphVisitor.GetCfgEdges(), graphVisitor.GetCdgEdges(),
                graphVisitor.getReachingDefEdges());
        }
    }

    public static List<string> ConvertToLines(string text)
    {
        string[] separators = { "\r\n", "\n" };
        return new List<string>(Regex.Split(text, string.Join("|", separators)));
    }
}

File: Preprocessing.Java/src/main/java/org/example/GraphGenerator.java
package org.example;

import com.github.javaparser.StaticJavaParser;
import com.github.javaparser.ast.CompilationUnit;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;

public class GraphGenerator {
    public static List<Graph> generate(Path sourceCodeFile) throws IOException {
        CompilationUnit cu = StaticJavaParser.parse(sourceCodeFile);

        MethodDeclarationVisitor methodVisitor = new MethodDeclarationVisitor();
        methodVisitor.visit(cu, null);

        var methods = methodVisitor.getMethods();
        var graphs = new ArrayList<Graph>();

        for (Method method : methods) {
            var className = method.classDeclaration().getNameAsString();
            var methodDeclaration = method.declaration();
            var methodName = String.format("%s->%s", className, methodDeclaration.getNameAsString());

            int lineNumber = methodDeclaration.getBegin().map(position -> position.line).orElse(-1);

            var lines = readLines(sourceCodeFile);

            var graphVisitor = new AstVisitor();
            graphVisitor.process(methodDeclaration);

            graphs.add(new Graph(sourceCodeFile.toString(), "unknown", methodName, lineNumber, 0,
                    lines, graphVisitor.getNodes(),
                    graphVisitor.getAstEdges(), graphVisitor.getCfgEdges(), graphVisitor.getCdgEdges(),
                    graphVisitor.getReachingDefEdges()));
        }

        return graphs;
    }

    public static List<String> readLines(Path filePath) throws IOException {
        return Files.readAllLines(filePath, StandardCharsets.UTF_8);
    }
}
